purchase_invoice.py

```diff
@@ -875,7 +884,188 @@ def parse_invoice(self,infile,account=None,paid_by_submitter=False,
+                for prop in line_item:
+                    if prop['type'] == 'line_item/description':
+                        s_item.description = prop['value']
+                    elif prop['type'] == 'line_item/product_code':
+                        s_item.item_code = prop['value']
+                    elif prop['type'] == 'line_item/quantity' and s_item.qty is None:
+                        s_item.qty = int(prop['value'])
+                    elif prop['type'] == 'line_item/unit' and s_item.qty_unit is None:
+                        s_item.qty_unit = prop['value']
+                    elif prop['type'] == 'line_item/amount':
+                        s_item.amount = float(prop['value'])
+                if s_item.description and "Vorkasse" in s_item.description:
+                    continue
+                if s_item.qty_unit and s_item.qty_unit == "Rol":
+                    try:
+                        r1 = re.search('[0-9]+ *[mM]', s_item.description)
+                        r2 = re.search('[0-9]+', r1.group(0))
+                        s_item.qty_unit = "Meter"
+                        s_item.qty = int(r2.group(0))
+                    except Exception:
+                        pass
+                if s_item.qty:
+                    s_item.rate = round(s_item.amount / s_item.qty, 2)
+                    rounding_error += s_item.amount - s_item.rate * s_item.qty
+                    self.items.append(s_item)
+        self.shipping += rounding_error
+        self.totals[self.default_vat] = float(
+            [el.get('value') for el in invoice_json['entities'] if el.get('type') == 'net_amount'][0])
+        self.vat[self.default_vat] = float(
+            [el.get('value') for el in invoice_json['entities'] if el.get('type') == 'total_tax_amount'][0])
+        self.compute_total()
+        return self
+
+    def parse_generic_json(self, invoice_json, default_account=None, paid_by_submitter=False, is_test=False):
+        amount = ""
+        self.parser = "generic"
+        self.extract_items = False
+        self.shipping = 0.0
+        self.date = [el.get('value') for el in invoice_json['entities'] if el.get('type') == 'invoice_date'][0]
+        self.order_id = [el.get('value') for el in invoice_json['entities'] if el.get('type') == 'purchase_order'][0]
+        self.gross_total = float([el.get('value') for el in invoice_json['entities'] if el.get('type') == 'total_amount'][0])
+        self.no = [el.get('value') for el in invoice_json['entities'] if el.get('type') == 'invoice_id'][0]
+        self.vat[self.default_vat] = float(
+            [el.get('value') for el in invoice_json['entities'] if el.get('type') == 'total_tax_amount'][0])
+        self.totals[self.default_vat] = float(
+            [el.get('value') for el in invoice_json['entities'] if el.get('type') == 'net_amount'][0])
+
+        if not is_test:
+            if self.check_if_present():
+                return self
+        else:
+            return self
+
```

Comment:
The following lines are duplicated code. Please factor this out.

Reply:
Fixed